### ISSUE TITLE
Provide a C++ project template

### ISSUE THREAD

**Issue by tristan0x** [BlueBrain/hpc-coding-conventions#32]

**Comment by tristan0x** [comment]
### Requirements
Complete this [Basalt project](https://github.com/tristan0x/basalt/issues?q=is%3Aopen+is%3Aissue+project%3Atristan0x%2Fbasalt%2F2) first 

### Specifications

Create a cookiecutter project inspired by basalt and an existing cookingcutter template for [open source python project](https://github.com/audreyr/cookiecutter-pypackage). Input parameters:

* team name: string
* email: string
* open_source: boolean
* project_slug
* python_bindings: boolean

**Comment by pramodk** [comment]
👍 

Some important missing items are related to parallel / distributed aspects. For example, for new software components (including coreneuron) what we will additionally need is:

* Logger support for parallel/mpi application (i.e. logging from multiple mpi ranks, [old discussion](https://bbpteam.epfl.ch/project/issues/browse/CNEUR-22))
* Error handling for parallel/mpi applications  (same as above)
* Cmake supporting MPI enabled build (e.g. some applications directly set `CC=mpicc` and others use `CC=gcc` and use `FindMPI.cmake` module to get link libraries)
* GPU enabled build support : 
  * Building C++ project together with CUDA code
  * Building C++ project with CUDA as well as OpenACC
* Parallel tests support
* Cross compile build support (not that important)
  * certain components like mod2c/nmodl are built for front-end node
  * rest of the code is built for compute node

**Comment by tristan0x** [comment]
Thank you for these additions Pramod. You are right, some of these items can be added to the project template afterward. But I think there are already some elements that can be part of the first version:

> * Logger support for parallel/mpi application (i.e. logging from multiple mpi ranks, [old discussion](https://bbpteam.epfl.ch/project/issues/browse/CNEUR-22))

One should investigate whether spdlog is MPI ready, or if a layer of top of spdlog to ease the logging experience in MPI programs already exists or should be developed by us.

> * Error handling for parallel/mpi applications  (same as above)

IMO this belongs to a GoodPractices document in the C++ coding conventions. Don't you think @pramodk ?

> * Cmake supporting MPI enabled build (e.g. some applications directly set `CC=mpicc` and others use `CC=gcc` and use `FindMPI.cmake` module to get link libraries)

I suggest to add an input parameter `use_mpi` in the project template in the initial version. I like the *FindMPI* approach as it provides more insights about the MPI setup (version, flags, ...) and runtime environment (like the path to mpiexec/mpirun) that are useful during the *test* phase.

> * GPU enabled build support :
>   
>   * Building C++ project together with CUDA code
>   * Building C++ project with CUDA as well as OpenACC

Maybe in a future version.

> * Parallel tests support

The CMake project that comes with hpc-coding-conventions should provide a bunch of CMake helpers like [the `add_mpi_test` function in zee](https://github.com/BlueBrain/zee/blob/master/CMakeLists.txt#L60-L94):

Usage:
```cmake
add_mpi_test(
    NAME LaplacianPETSc
    NUM_PROCS 16
    COMMAND $<TARGET_FILE:LaplacianPETSc>)
```

add_mpi_test(NUM_PROC 16 COMMAND

> * Cross compile build support (not that important)
>   
>   * certain components like mod2c/nmodl are built for front-end node
>   * rest of the code is built for compute node

This should be addressed later also.

**Comment by pramodk** [comment]
> One should investigate whether spdlog is MPI ready, or if a layer of top of spdlog to ease the logging experience in MPI programs already exists or should be developed by us.

yes, small layer on spdlog should do the job. (I dont think spdlog is mpi ready).

> IMO this belongs to a GoodPractices document in the C++ coding conventions. Don't you think  @pramodk ?

What I meant us some kind of common utility classes  for logging and error handling. For example, in MPI application, we typically do following throughout the application: 

```cpp
if (rank == 0) {
    logger->error(...);
    abort(); // MPI_Abort();
}
```
Some common way would be useful for multiple applications.
